x-pack/test/functional_enterprise_search/services/app_search_client.ts

```diff
@@ -0,0 +1,114 @@
+ */
+const makeRequest = <T>(method: string, path: string, body?: object): Promise<T> => {
+  return new Promise(function(resolve, reject) {
+    // Throw error if no key is configured
+    // Handle API errors
+    const APP_SEARCH_API_KEY = process.env.APP_SEARCH_API_KEY || '';
+    let postData;
+
+    if (body) {
+      postData = JSON.stringify(body);
+    }
+
+    const req = http.request(
+      {
+        method,
+        hostname: 'localhost',
+        port: 3002,
+        path,
+        agent: false, // Create a new agent just for this one request
+        headers: {
+          Authorization: `Bearer ${APP_SEARCH_API_KEY}`,
+          'Content-Type': 'application/json',
+          ...(!!postData && { 'Content-Length': Buffer.byteLength(postData) }),
+        },
+      },
+      res => {
+        const bodyChunks: Uint8Array[] = [];
+        res.on('data', function(chunk) {
+          bodyChunks.push(chunk);
+        });
+
+        res.on('end', function() {
+          let responseBody;
+          try {
+            responseBody = JSON.parse(Buffer.concat(bodyChunks).toString());
+          } catch (e) {
+            reject(e);
+          }
+
+          resolve(responseBody);
+        });
+      }
+    );
+
+    req.on('error', e => {
+      reject(e);
+    });
+
+    if (postData) {
+      req.write(postData);
+    }
+    req.end();
+  });
+};
+
+export interface IEngine {
+  name: string;
+}
+
+export const createEngine = async (engineName: string): Promise<IEngine> => {
+  return await makeRequest('POST', '/api/as/v1/engines', { name: engineName });
+};
+
+export const destroyEngine = async (engineName: string): Promise<object> => {
+  return await makeRequest('DELETE', `/api/as/v1/engines/${engineName}`);
+};
+
+export const createMetaEngine = async (
+  engineName: string,
+  sourceEngines: string[]
+): Promise<IEngine> => {
+  return await makeRequest('POST', '/api/as/v1/engines', {
+    name: engineName,
+    type: 'meta',
+    source_engines: sourceEngines,
+  });
+};
+
+export interface ISearchResponse {
+  results: object[];
+}
+
+const search = async (engineName: string): Promise<ISearchResponse> => {
+  return await makeRequest('POST', `/api/as/v1/engines/${engineName}/search`, { query: '' });
+};
+
+// Since the App Search API does not issue document receipts, the only way to tell whether or not documents
+// are fully indexed is to poll the search endpoint.
+export const waitForIndexedDocs = (engineName: string) => {
+  return new Promise(async function(resolve) {
+    let isReady = false;
+    while (!isReady) {
+      const response = await search(engineName);
+      if (response.results && response.results.length > 0) {
+        isReady = true;
+        resolve();
+      }
+    }
+  });
+};
+
+export const indexData = async (engineName: string, docs: object[]) => {
+  return await makeRequest('POST', `/api/as/v1/engines/${engineName}/documents`, docs);
+};
```

Comment:
Just wanted to say this is a super impressive client - huge kudos! 👏 

Do you envision that we'll be able to potentially phase it out at some point and be able to complete all API actions within the Kibana UI naturally? (kind of the same way we do right now for Cypress tests in AS?)

Reply:
I think if we ever had the ability to create engines from within the UI, then yes, then I would imagine we could phase this out.

I also wondered about using the `esArchiver` to load fixtures into ES for our various tests, which would be pretty neat.